Filter Tag Manager containers by usage context

aaemnnosttv opened the pull request:
## Summary

Addresses issue #470 #591 #821

#### With AMP in Transitional / Off

![image](https://user-images.githubusercontent.com/1621608/68512080-0b7c8200-0280-11ea-9efe-5dc6495ec12b.png)

#### With AMP in Standard / Primary mode

![image](https://user-images.githubusercontent.com/1621608/68510501-644a1b80-027c-11ea-9f8b-a465c3cfa2b2.png)

## Relevant technical choices

- Discovered and addresses #821 
- Addresses #591 to avoid merge conflicts and extra view later
- Simplifies some of the logic and verbosity in the tag manager setup component

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

felixarntz commented on assets/js/modules/tagmanager/setup.js:
```diff
@@ -417,16 +414,16 @@ class TagmanagerSetup extends Component {
-								</Option> ) }
+									key={ i }
+									value={ publicId /* Capitalization rule exception: `publicId` is a property of an API returned value. */ }
+									children={ name }
```
I feel like it's a bit weird passing child elements this way, would prefer if we consistently did this like typical JSX, as a child node. But not blocking.

aaemnnosttv commented:
> Code-wise this mostly looks good, I fixed a couple minor issues I found.

I only saw the one change for ensuring the option was an array. We may want to consider adding a `default` parameter to `Options::get()` for this reason. Without it we need to use `Options::get( <name> ) ?: default` otherwise casting it to an array will result in `array( false )`. I've updated that line accordingly.

> Filtering containers works as expected in both AMP and non-AMP. However, I was not able to create a new GTM container with these changes. In both AMP and non-AMP, I received the following 500 response:

I was unable to reproduce this. I'm rather surprised that this happened as well as I tested creating containers many times with both. Can you share more details? Is there a fatal error in your logs?

I did find a minor change that was left to be done regarding an update to the `connection` endpoint as this is used to populate the module settings passed to JS. This did not yet contain the `ampContainerID` key so the settings were showing the wrong container ID after switching to a primary AMP mode.

felixarntz commented:
@aaemnnosttv 
> We may want to consider adding a `default` parameter to `Options::get()` for this reason.

Better handling defaults is a great idea, however I'd prefer if we did this through WordPress's API function `register_setting()`, which makes it more sustainable. The fact that you have to pass default values in WP every time you call `get_option()` isn't great because it easily allows for mistakes of using different defaults, or accidentally omitting it in one place. I think it'd be great to register every setting that Site Kit uses in the long run with proper configuration. How about we create an issue for now to do that for specifically all module settings (which mostly are arrays like that)? We could review all our regular settings later.

felixarntz commented on assets/js/modules/tagmanager/setup.js:
```diff
@@ -420,19 +417,16 @@ class TagmanagerSetup extends Component {
+								<Option
+									key={ i }
+									value={ publicId /* Capitalization rule exception: `publicId` is a property of an API returned value. */ }
+									children={ name }
```
Why do it this way and not like commonly done in React with proper child nesting `<Option ...>{ name }</Option>`?

aaemnnosttv replied:
This is just a stylistic preference which seemed a bit cleaner at the time, particularly since the component is spread vertically. If it were `<Option>value</Option>` without other props I definitely would have written it that way. Happy to change this if you'd like.

felixarntz replied:
I think for consistency we should always pass children as proper child-like nodes. Will update it accordingly quickly.

felixarntz commented:
@aaemnnosttv 
> I was unable to reproduce this. I'm rather surprised that this happened as well as I tested creating containers many times with both. Can you share more details? Is there a fatal error in your logs?

The problem is my site title, which includes an `&amp;` (since WP saves them in the DB with encoded entities like that). We need to do two things:
* Use `wp_specialchars_decode()` on the site title (keep `remove_accents()` as well).
* We need to figure out how we should deal with special characters like `&`. Maybe we should just strip them. In theory, it'd be awesome if we were as intelligent as replacing `&` with `and` for example, but that would either be super-specific or would need to cover similar patterns too which would make it a bit overkill.

aaemnnosttv commented:
> The problem is my site title, which includes an &amp; (since WP saves them in the DB with encoded entities like that).

I see what you mean now. When adding an ampersand I was able to see the same error, which was just Tag Manager rejecting the request since it contained "reserved characters".

I wasn't able to find any documentation around which characters are allowed, but the web interface does do some on-the-fly validation which I used to test a few things. Interestingly, accent characters were not flagged as invalid.

![image](https://user-images.githubusercontent.com/1621608/68846923-7988d500-06d6-11ea-8540-dd55a92fdd2b.png)

I ended up keeping `remove_accents` though as it was the safest way to be able to remove everything else that we know was invalid.

I also added a fallback for the container name to use the domain of the reference URL, as we've seen some sites have empty blog names cause problems before.

Finally, I extracted this sanitization to a new helper method and added a test for it to make sure this is catching all the things.

Back to you @felixarntz
